### PR TITLE
Constantin/dao/sales

### PR DESCRIPTION
**Pull request by DjConsti**
Bitte reviewen!

### CODE REVIEW

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Hashtable;
+import java.util.List;
+
+/**
+ * This is a helper class for the H2 Database implementations DBIntermittentSaleDAO and DBOnetimeSaleDAO
+ */
+public class DBAbstractSaleDAO {
```
Comment (PhilippParis) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: public abstract class

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Sale.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @return true if the Sale is active at the given time
+     */
+    public abstract boolean isAt(LocalDateTime time);
+
+
+    /**
+     *
+     * @return the entries of the sale with the new price
+     */
+    public Hashtable<MenuEntry, BigDecimal> getEntries() {
```
Comment (PhilippParis) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Sale.java]: maybe:

public Set<MenuEntry> getMenuEntries() {
return entries.keySet();
}

public BigDecimal getSalePriceOf(MenuEntry entry) {
return entries.get(entry);
}

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+    @Autowired
+    private DataSource dataSource;
+    @Autowired
+    private DAO<MenuEntry> menuEntryDAO;
+
+    @Override
+    public void create(T sale) throws DAOException, ValidationException {
+        //Save Sale
+        final String saleQuery = "INSERT INTO Sale (name, deleted) VALUES (?, ?)";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: remove deleted

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(saleQuery)) {
+            stmt.setString(1, sale.getName());
+            stmt.setBoolean(2, false);
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: remove (default value of deleted is false)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    /**
+     * writes the association changes of the dataset into the database
+     * stores the time; number of the change (uses the given changeNumber) and the user which executed the changes
+     * @param sale updated dataset
+     * @param changeNumber reservation history change number
+     * @throws DAOException if an error accessing the database occurred
+     */
+    private void generateReservationAssociationsHistory(Sale sale, long changeNumber) throws DAOException {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: reservation in sale?

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        LOGGER.debug("Entering generateReservationAssociationsHistory with parameters: " + sale);
+
+        final String query = "INSERT INTO SaleAssocHistory " +
+                "(SELECT *, CURRENT_TIMESTAMP(), ?, ? FROM SaleAssoc WHERE sale_ID = ?)";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: attributes are missing and assoc key is wrong (should be sale_id + entry_id)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    /**
+     * writes the changes of the dataset into the database
+     * stores the time; number of the change and the user which executed the changes
+     * @param sale updated dataset
+     * @throws DAOException if an error accessing the database occurred
+     */
+    protected void generateSaleHistory(Sale sale) throws DAOException {
+        LOGGER.debug("Entering generateHistory with parameters: " + sale);
+
+        final String query = "INSERT INTO SaleHistory " +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: attributes are missing

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.List;
+
+/**
+ * This is a helper class for the H2 Database implementations DBIntermittentSaleDAO and DBOnetimeSaleDAO
+ */
+abstract class DBAbstractSaleDAO<T extends Sale> implements DAO<T> {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: add @PreAuthorize("isAuthenticated()")

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+            sale.setName(name);
+        } catch (SQLException e) {
+            LOGGER.error("reading from SALE table failed", e);
+            throw new DAOException("reading from SALE table failed", e);
+        }
+
+        // set entries
+        // SaleAssoc (sale_ID, menuEntry_ID, salePrice, disabled)
+
+        final String entriesQuery = "SELECT * FROM SaleAssoc WHERE sale_ID = ? AND disabled = false";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: wrong id

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(entriesQuery)) {
+            stmt.setLong(1, sale.getIdentity());
+            ResultSet result = stmt.executeQuery();
+
+            List<MenuEntry> entries = new ArrayList<>();
+
+            while(result.next()) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: replace it by:

while (result.next()) {
   entries.add(MenuEntry.withIdentity(result.getLong("menuEntry_ID")));
}
menuEntryDAO.populate(entries);

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+
+        } catch (SQLException e) {
+            LOGGER.error("reading from SaleHistory table failed", e);
+            throw new DAOException("reading from SaleHistory table failed", e);
+        }
+
+        // set entries
+        // SaleAssoc (sale_ID, menuEntry_ID, salePrice, disabled)
+
+        final String entriesQuery = "SELECT * FROM SaleAssocHistory WHERE sale_ID = ? AND changeNr = ? AND disabled = false";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: wrong id

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        try (PreparedStatement stmt = dataSource.getConnection().prepareStatement(entriesQuery)) {
+            stmt.setLong(1, history.getData().getIdentity());
+            stmt.setLong(2, history.getChangeNumber());
+            ResultSet result = stmt.executeQuery();
+
+            List<MenuEntry> entries = new ArrayList<>();
+
+            while(result.next()) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBAbstractSaleDAO.java]: replace it by:

while (result.next()) {
entries.add(MenuEntry.withIdentity(result.getLong("menuEntry_ID")));
}
menuEntryDAO.populate(entries);

and update the prices afterwards (have a look at MenuDAO how this should be done)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+
+    @Override
+    public List<IntermittentSale> find(IntermittentSale intermittentSale) throws DAOException {
+        LOGGER.debug("Entering find with parameters: " + intermittentSale);
+
+        if (intermittentSale == null) {
+            return new ArrayList<>();
+        }
+
+        final String query = "SELECT * FROM IntermittentSale WHERE sale_ID = ISNULL(?, sale_ID)" +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java]: natural join IntermittentSale <-> Sale

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+        }
+
+        return intermittentSales;
+    }
+
+    @Override
+    public List<IntermittentSale> getAll() throws DAOException {
+        LOGGER.debug("Entering getAll");
+
+        final String query = "SELECT * FROM IntermittentSale WHERE EXISTS (SELECT * FROM Sale WHERE IntermittentSale.sale_ID = Sale.ID AND deleted = false)";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java]: natural join

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+        if (intermittentSales == null || intermittentSales.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        for (IntermittentSale order : intermittentSales) {
+            validator.validateIdentity(order);
+        }
+
+        final String query = "SELECT * FROM IntermittentSale WHERE sale_ID IN (" +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java]: natural join sale

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+    /**
+     * converts the database query output into a history entry object
+     * @param result database output
+     * @return History object with the data of the resultSet set
+     * @throws SQLException if an error accessing the database occurred
+     * @throws DAOException if an error retrieving the sale ocurred
+     */
+    private History<IntermittentSale> historyFromResultSet(ResultSet result) throws DAOException, SQLException {
+        // get user
+        List<User> storedUsers = userDAO.find(User.withIdentity(result.getString("changeUser")));
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBIntermittentSaleDAO.java]: populate

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+
+    @Override
+    public List<OnetimeSale> find(OnetimeSale onetimeSale) throws DAOException {
+        LOGGER.debug("Entering find with parameters: " + onetimeSale);
+
+        if (onetimeSale == null) {
+            return new ArrayList<>();
+        }
+
+        final String query = "SELECT * FROM OnetimeSale WHERE sale_ID = ISNULL(?, sale_ID)" +
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java]: join

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+        }
+
+        return onetimeSales;
+    }
+
+    @Override
+    public List<OnetimeSale> getAll() throws DAOException {
+        LOGGER.debug("Entering getAll");
+
+        final String query = "SELECT * FROM OnetimeSale WHERE EXISTS (SELECT * FROM Sale WHERE OnetimeSale.sale_ID = Sale.ID AND deleted = false)";
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java]: join

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+    /**
+     * converts the database query output into a history entry object
+     * @param result database output
+     * @return History object with the data of the resultSet set
+     * @throws SQLException if an error accessing the database occurred
+     * @throws DAOException if an error retrieving the sale ocurred
+     */
+    private History<OnetimeSale> historyFromResultSet(ResultSet result) throws DAOException, SQLException {
+        // get user
+        List<User> storedUsers = userDAO.find(User.withIdentity(result.getString("changeUser")));
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/impl/DBOnetimeSaleDAO.java]: populate

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/service/SaleService.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.history.History;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.ValidationException;
+import org.springframework.security.access.prepost.PreAuthorize;
+
+import java.util.List;
+
+/**
+ * Service interface for sales.
+ */
+public interface SaleService {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/service/SaleService.java]: extends Service

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/service/impl/SaleServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.List;
+
+/**
+ * Implementation of the SaleService
+ */
+public class SaleServiceImpl implements SaleService {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/service/impl/SaleServiceImpl.java]: remove public

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/IntermittentSaleValidator.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
+package com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.impl;
+
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.IntermittentSale;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.ValidationException;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.Validator;
+
+/**
+ * validator implementation for the IntermittentSale domain object
+ */
+public class IntermittentSaleValidator implements Validator<IntermittentSale> {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/IntermittentSaleValidator.java]: remove public

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java [hunk 23]
```diff
@@ -0,0 +1,74 @@
+package com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.impl;
+
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.OnetimeSale;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.ValidationException;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.Validator;
+
+/**
+ * validator implementation for the OnetimeSale domain object
+ */
+public class OnetimeSaleValidator implements Validator<OnetimeSale> {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java]: remove public

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        if (onetimeSale.getName() == null) {
+            throw new ValidationException("name must not be null");
+        }
+        if (onetimeSale.getName().isEmpty()) {
+            throw new ValidationException("name must not be empty");
+        }
+        if (onetimeSale.getEntries() == null){
+            throw new ValidationException("Menu entries not be null");
+        }
+        if (onetimeSale.getEntries().size() == 0){
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java]: remove

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+        if (onetimeSale.getFromTime() == null) {
+            throw new ValidationException("FromTime must not be null");
+        }
+        if (onetimeSale.getToTime() == null) {
+            throw new ValidationException("ToTime must not be null");
+        }
+        if (onetimeSale.getFromTime().isAfter(onetimeSale.getToTime())) {
+            throw new ValidationException("FromTime must not be after toTime.");
+        }
+    }
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/OnetimeSaleValidator.java]: entries validation missing (validateIdentity of entries)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/IntermittentSaleValidator.java [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+        if(intermittentSale.getDaysOfSale() == null) {
+            throw new ValidationException("DaysOfSale must not be null");
+        }
+        if(intermittentSale.getDaysOfSale().isEmpty()) {
+            throw new ValidationException("There should be at least one day of sale");
+        }
+        if (intermittentSale.getDuration() == null || intermittentSale.getDuration().isNegative()) {
+            throw new ValidationException("Negative durations are not accepted.");
+        }
+    }
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/IntermittentSaleValidator.java]: entries validation missing (validateIdentity of entries)

**Pull request merged**